README
=========================
Procrastinate No More 0.2
=========================

Description
===========

Procrastinate No More is a simple widget for the KDE Plasma Workspaces to help you avoid 
procrastination. It is inspired by a Python applet called Tomate, which you can read more
about here: https://gitorious.org/tomate/pages/Home

Have you ever thought "_now_ I will start doing what I am supposed to do", only to find 
yourself doing exactly everything except that? The reason we procrastinate is often not
because of the task itself, the problem is to get started with it. Once we have passed
that stage, however, it is usually easy to just go with the flow and get work done.

Procrastinate No More can help you get past that first stage. When you click on the
widget, you make a contract with yourself to commit fully to a task for 10 minutes without 
any distractions. After 10 minutes have passed you are free to decide whether you want to 
continue working or do something else.

Why 10 minutes? It is short enough to always be doable, no matter how boring or daunting
the task might seem, while it should let you stay concentrated long enough to pick up the
flow.

Installation
============

If possible, it is recommended that you install this widget through your package manager.
The typical procedure to compile it from source is as follows:

mkdir build
cd build
cmake -DCMAKE_INSTALL_PREFIX=`kde4-config --prefix` ..
make
make install

The last command usually requires root permissions so you probably need to use su/sudo
or similar commands.

How to use
==========

The widget can be added to the desktop, a panel or to the system tray.

1. When you want to commit yourself to a task, click on the widget. The widget becomes red.
2. Work.
3. After 10 minutes the widget turns green. Now you can either continue to work, or do
something else.
4. When you stop working, regardless if the 10 minutes have passed or not, click on the 
widget again. It is now back to its idle state (blue icon) and the tooltip tells you for
how long you worked.
5. Rinse and repeat.

Configuration options
=====================

Procrastinate No More follows the path of its inspiration (Tomate) and tries to be as
simple as possible. For that reason it only has the most essential features and no
options you can waste time tweaking.

For those who absolutely have to change the minimum working time or the icons that 
represent the different states, see the short description below.

Change minimum working time
===========================

The minimum working time (10 minutes by default) is defined in src/tomat.h on line 46:

    enum { MIN_WORK_TIME = 10*60 }; // minimum working time in seconds

Change 10*60 to whatever time you want (given in seconds). It is necessary to recompile
the widget after making the change.

Change state icons
==================

The icons can be defined by the Plasma Workspace theme (widgets/work_state.svgz).
The default icons are located in the file

    <kde prefix>/share/apps/desktoptheme/default/widgets/work_state.svgz

where <kde prefix> can be found with the command `kde4-config --prefix` (usually /usr).
The SVG file contains three icons with the following id for the different states:

idle
working
ok

Trivia
======

- Internally Procrastinate No More goes by the name "Tomat", Swedish for tomato. The name
comes from the applet it is inspired by, Tomate, which in turn is named that way because
the concept is similar to the popular Pomodoro technique. ("Pomodoro" is tomato in
Italian.)
- The author started coding this widget just before one of his final exams.



CHANGELOG
0.2
---
- Add support for keyboard shortcut

NOTE: If the widget is in the system tray, the shortcut is not saved between
logins due to a bug (https://bugs.kde.org/show_bug.cgi?id=260685)

0.1
---
- Initial release


src/tomat.h
/***************************************************************************
 * Copyright © 2011 Hans Chen <contact@hanschen.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the
 * Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 ***************************************************************************/

#ifndef TOMAT_H
#define TOMAT_H

#include <QTimer>
#include <QString>

#include <Plasma/Applet>
#include <Plasma/Svg>

enum State {IDLE, WORKING, OK};

class Tomat : public Plasma::Applet
{
    Q_OBJECT

    public:
        Tomat(QObject *parent, const QVariantList &args);
        ~Tomat();

        void init();
        void paintInterface(QPainter *p,
                            const QStyleOptionGraphicsItem *option,
                            const QRect &contentsRect);
        void constraintsEvent(Plasma::Constraints constrants);
        void mousePressEvent(QGraphicsSceneMouseEvent *event);

        enum { MIN_WORK_TIME = 10*60 }; // minimum working time in seconds

    public slots:
        void toolTipAboutToShow();
        void toolTipHidden();

    protected:
        QString getTime();
        QString getToolTip();

    private slots:
        void setState(State newState);
        void manuallySwitchState();
        void resetToolTip();
        void updateTime();
        void reloadTheme();

    private:
        QTimer workingTimer;
        Plasma::Svg svg;
        State currentState; 
        State oldState; 
        int secondsWorking;
};

#endif



src/tomat.cpp
/***************************************************************************
* Copyright © 2011 Hans Chen <contact@hanschen.org>
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation; either version 2 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the
* Free Software Foundation, Inc.,
* 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
***************************************************************************/

#include "tomat.h"

#include <QtCore/QSizeF>
#include <QtCore/QTimer>
#include <QtGui/QGraphicsSceneMouseEvent>
#include <QtGui/QPainter>

#include <Plasma/Svg>
#include <Plasma/Theme>
#include <Plasma/ToolTipContent>
#include <Plasma/ToolTipManager>

Tomat::Tomat(QObject *parent, const QVariantList &args)
    : Plasma::Applet(parent, args),
      svg(this)
{
    currentState = IDLE;
    oldState = IDLE;

    svg.setImagePath(Plasma::Theme::defaultTheme()->imagePath("widgets/work_state"));
    svg.setContainsMultipleImages(true);

    setAspectRatioMode(Plasma::Square);
    setBackgroundHints(DefaultBackground);
}

Tomat::~Tomat()
{
}

void Tomat::init()
{
    Plasma::ToolTipManager::self()->registerWidget(this);

    connect(&workingTimer, SIGNAL(timeout()), this, SLOT(updateTime()));
    connect(Plasma::Theme::defaultTheme(), SIGNAL(themeChanged()), this, SLOT(reloadTheme()));
    connect(this, SIGNAL(activate()), this, SLOT(manuallySwitchState()));
}

void Tomat::paintInterface(QPainter *p,
                           const QStyleOptionGraphicsItem *option,
                           const QRect &contentsRect)
{
    const double margins = 4.0; // thickness of margins around icon in px

    Q_UNUSED(option)
    
    p->setRenderHint(QPainter::SmoothPixmapTransform);
    p->setRenderHint(QPainter::Antialiasing);

    double width = contentsRect.width(),
           height = contentsRect.height();

    QString elementid;
    switch (currentState) {
        case IDLE:
            elementid = "idle";
            break;
        case WORKING:
            elementid = "working";
            break;
        case OK:
            elementid = "ok";
            break;
    }

    if (width < height) {
        height = width;
    } else {
        width = height;
    }

    p->save();

    svg.resize(QSizeF(width - margins, height - margins));
    svg.paint(p,
             (int)contentsRect.left() + margins/2,
             (int)contentsRect.top() + margins/2,
             elementid);

    p->restore();
}

void Tomat::constraintsEvent(Plasma::Constraints constraints)
{
    if (constraints & Plasma::FormFactorConstraint) {
        if (formFactor() == Plasma::Horizontal || formFactor() == Plasma::Vertical) {
            setPreferredSize(-1,-1);
        } else {
            setPreferredSize(100, 100);
        }

    }

    if (formFactor() == Plasma::Planar || formFactor() == Plasma::MediaCenter) {
        setMinimumSize(32,32);
    } else {
        setMinimumSize(0,0);
    }

    update();
}

void Tomat::mousePressEvent(QGraphicsSceneMouseEvent *event)
{

    if (event->buttons() == Qt::LeftButton &&
        contentsRect().contains(event->pos())) {
        manuallySwitchState();
        return;
    }

    Applet::mousePressEvent(event);
}

void Tomat::toolTipAboutToShow()
{
    Plasma::ToolTipContent toolTip;
    toolTip.setMainText(getToolTip());
    Plasma::ToolTipManager::self()->setContent(this, toolTip);
}

void Tomat::toolTipHidden()
{
}

void Tomat::setState(State newState)
{
    if (newState == IDLE) {
        workingTimer.stop();
        QTimer::singleShot(60*1000, this, SLOT(resetToolTip()));
        Plasma::ToolTipManager::self()->show(this);
    }
    else if (newState == WORKING) {
        secondsWorking = 0;
        workingTimer.start(1000);
    }

    oldState = currentState;
    currentState = newState;
    update();
}

void Tomat::manuallySwitchState()
{
    if (currentState == IDLE) {
        setState(WORKING);
    } else {
        setState(IDLE);
    }
}

QString Tomat::getTime()
{
    const int secondsInMinute = 60;
    const int secondsInHour = 60*60;

    QString time;

    int hours = secondsWorking/secondsInHour;
    int minutes = (secondsWorking%secondsInHour)/secondsInMinute;

    QString hourText = (hours == 1) ? i18n("hour")
                                    : i18n("hours");
    QString minuteText = (minutes == 1) ? i18n("minute")
                                        : i18n("minutes");

    if (hours > 0) {
        time = i18n("%1 %2 and ").arg(QString::number(hours))
                                 .arg(hourText);
    }
    time = time + i18n("%1 %2").arg(QString::number(minutes))
                               .arg(minuteText);
        
    return time;
}

QString Tomat::getToolTip()
{
    QString toolTip;

    if (currentState == IDLE) {
        if (oldState == IDLE) {
            toolTip = "Idle";
        }
        else if (oldState == WORKING) {
            toolTip = i18n("Not good, only worked for %1").arg(getTime());
        }
        else if (oldState == OK) {
            toolTip = i18n("Good! Worked for %1").arg(getTime());
        }
    } else {
        toolTip = i18n("Working for %1").arg(getTime());
    }

    return toolTip;
}

void Tomat::resetToolTip()
{
    // resets the tool tip to "Idle"
    if (currentState == IDLE) {
        oldState = IDLE;
    }
}

void Tomat::updateTime()
{
    secondsWorking++;
    if (currentState == WORKING && secondsWorking >= MIN_WORK_TIME) {
        setState(OK);
    }
}

void Tomat::reloadTheme()
{
    svg.setImagePath(Plasma::Theme::defaultTheme()->imagePath("widgets/work_state"));
}

K_EXPORT_PLASMA_APPLET(tomat, Tomat)
